stunnel/stunnel_cache.ml

```diff
@@ -23,40 +23,57 @@
+(* Disable debug-level logging but leave higher-priority enabled.  It would be
+ * better to handle this sort of configuration in the Debug module itself.
+ *)
+let debug_enabled = false
+
+let ignore_log fmt =
+    Printf.ksprintf (fun s -> ()) fmt
+
+(* Use and overlay the definition from D. *)
+let debug = if debug_enabled then debug else ignore_log
+
+type endpoint = { host: string; port: int; verified: bool }
+
+(* Need to limit the absolute number of stunnels as well as the maximum age *)
+let max_stunnel = 22
+let max_age = 180. *. 60. (* seconds *)
+let max_idle = 5. *. 60. (* seconds *)
+
+(* The add function adds the new stunnel before doing gc, so the cache *)
+(* can briefly contain one more than maximum. *)
+let capacity = max_stunnel + 1
 
 (** An index of endpoints to stunnel IDs *)
-let index : (endpoint, int list) Hashtbl.t ref = ref (Hashtbl.create 10)
+let index : (endpoint, int list) Hashtbl.t ref = ref (Hashtbl.create capacity)
 (** A mapping of stunnel unique IDs to donation times *)
-let times : (int, float) Hashtbl.t ref = ref (Hashtbl.create 10)
+let times : (int, float) Hashtbl.t ref = ref (Hashtbl.create capacity)
 (** A mapping of stunnel unique ID to Stunnel.t *)
-let stunnels : (int, Stunnel.t) Hashtbl.t ref = ref (Hashtbl.create 10)
+let stunnels : (int, Stunnel.t) Hashtbl.t ref = ref (Hashtbl.create capacity)
 
 open Pervasiveext
 open Threadext
 open Listext
 
 let m = Mutex.create ()
 
-(* Need to limit the absolute number of stunnels as well as the maximum age *)
-let max_stunnel = 4
-let max_age = 5. *. 60. (* seconds *)
-let max_idle = 2. *. 60. (* seconds *)
-
 let id_of_stunnel stunnel = 
     Opt.default "unknown" (Opt.map string_of_int stunnel.Stunnel.unique_id)
 
 let unlocked_gc () = 
-  let now = Unix.gettimeofday () in
-  let string_of_id id = 
-    let stunnel = Hashtbl.find !stunnels id in
-    Printf.sprintf "(id %s / idle %.2f age %.2f)" 
-      (id_of_stunnel stunnel)
-      (now -. (Hashtbl.find !times id))
-      (now -. stunnel.Stunnel.connected_time) in
-  let string_of_endpoint ep = Printf.sprintf "%s:%d" ep.host ep.port in
-  let string_of_index ep xs = Printf.sprintf "[ %s %s ]" (string_of_endpoint ep) (String.concat "; " (List.map string_of_id xs)) in
-  debug "Cache contents: %s" (Hashtbl.fold (fun ep xs acc -> string_of_index ep xs ^ " " ^ acc) !index "");
+	if debug_enabled then begin
+		let now = Unix.gettimeofday () in
+		let string_of_id id = 
+			let stunnel = Hashtbl.find !stunnels id in
+			Printf.sprintf "(id %s / idle %.2f age %.2f)" 
+				(id_of_stunnel stunnel)
+				(now -. (Hashtbl.find !times id))
+				(now -. stunnel.Stunnel.connected_time) in
+		let string_of_endpoint ep = Printf.sprintf "%s:%d" ep.host ep.port in
+		let string_of_index ep xs = Printf.sprintf "[ %s %s ]" (string_of_endpoint ep) (String.concat "; " (List.map string_of_id xs)) in
+		debug "Cache contents: %s" (Hashtbl.fold (fun ep xs acc -> string_of_index ep xs ^ " " ^ acc) !index "");
```

Comment:
Is `Hashbl.fold` here used to construct a long string? If so, it would be better to use `String.concat` as string concatenation with `^` is not efficient. If the string is really long, using `Buffer.t` would be even better.